### ISSUE TITLE
Make if lazy

### ISSUE THREAD

**Issue by mariari** [heliaxdev/minijuvix#217]
Currently if is defined like

```haskell
  if : {A : Type} → Bool → A → A → A;
  if true a _ ≔ a;
  if false _ b ≔ b;
```

and since the language is not lazy, this means both `a` and `b` will be evaluated.

This could be fixed by making `if` a "macro" or a "special form" in which it has lazy semantics. Likewise `||` and `&&` should have the same properties (though `||` and `&&` are special cases of `if` and thus not "special forms").

Further functions like:

```haskell
ifOrd : {A : Type} → Ordering -> A → A → A → A;
ifOrd LT lt eq gt ≔ lt;
ifOrd EQ lt eq gt ≔ eq;
ifOrd GT lt eq gt ≔ gt;
```

should be deprecated, or a macro system should be had if this kind of abstraction is wanted. Or Alternatively a thunk version should be had if such form is wanted but no macro is had.

```haskell
ifOrd : {A : Type} → Ordering -> (Unit -> A) → (Unit -> A) → (Unit -> A) → A;
ifOrd LT lt eq gt ≔ lt unit;
ifOrd EQ lt eq gt ≔ eq unit;
ifOrd GT lt eq gt ≔ gt unit;
```

**Comment by janmasrovira** [comment]
We want to investigate the possibility to support both call-by-value and call-by-name. However, it is not clear yet if we'll focus on this soon or not.

**Comment by mariari** [comment]
That seems quite ambitious, from what I understand Haskell has a lot of machinery (STG machine) in the compilation to make `call-by-name` fast. Thus it'll probably have a non noticeable affect on the binary size and speed and complexity of writing the compiler? A simpler approach is to either have special forms like that, or have a basic macro system that you can tell to auto box functions in lambda

```haskell
Inductive Thunk A {
  box : Unit → A;
};

call : Thunk A → A;
call (box thunk) ≔ thunk unit;

if-func : {A : Type} → Bool → (Thunk A) → (Thunk A) → A;
if-func true a _ ≔ call a;
if-func false _ b ≔ call b;

macro if : {A : Type} → Bool → A → A → A;
macro if x y z = `(if-func x (box ,y) (box ,z))
```

This approach still has costs if you don't make if special (I.E. implement something akin to the `if` above), namely the boxing of the arguments in `if` meaning an allocation is costed upon the calling of `if` which is something that would ideally be avoided.

**Comment by lukaszcz** [comment]
Indeed, closures are the way to implement laziness in an eager functional programming language. The OCaml compiler does exactly that under the hood. This works fine if you're eager most of the time and need laziness only occasionally. Though then one definitely needs "if", "&&", "||", etc., as special syntactic forms which are compiled separately. Or one can implement them with a macro system, but having a macro system just for this is overkill. Whether we need a macro system in general is a longer and separate discussion.
